ecommerce/extensions/payment/processors/payu.py

```diff
@@ -0,0 +1,292 @@
+                '6': TransactionDeclined,
+                '104': GatewayError
+            }.get(decision, InvalidPayUDecision)
+
+            raise exception
+
+        # Create Source to track all transactions related to this processor and order
+        source_type, __ = SourceType.objects.get_or_create(name=self.NAME)
+        currency = response['currency']
+        total = Decimal(response['value'])
+        transaction_id = response['transaction_id']
+        req_card_number = response['cc_number'] or ''
+        # card_type = CYBERSOURCE_CARD_TYPE_MAP.get(response['req_card_type'])
+
+        source = Source(source_type=source_type,
+                        currency=currency,
+                        amount_allocated=total,
+                        amount_debited=total,
+                        reference=transaction_id,
+                        label=req_card_number)
+
+        # Create PaymentEvent to track
+        event_type, __ = PaymentEventType.objects.get_or_create(name=PaymentEventTypeName.PAID)
+        event = PaymentEvent(event_type=event_type, amount=total, reference=transaction_id, processor_name=self.NAME)
+
+        return source, event
+
+    def _generate_signature(self, parameters, signature_type=1):
+        """
+        Sign the contents of the provided transaction parameters dictionary.
+
+        This allows PayU to verify that the transaction parameters have not been tampered with
+        during transit.
+
+        We also use this signature to verify that the signature we get back from PayU is valid for
+        the parameters that they are giving to us.
+
+        Arguments:
+            parameters (dict): A dictionary of transaction parameters.
+
+        Returns:
+            unicode: the signature for the given parameters
+        """
+
+        # sigantures to validate payment form
+        if signature_type == 1:
+            uncoded = "{ApiKey}~{merchantId}~{referenceCode}~{amount}~{currency}".format(
+                ApiKey=self.ApiKey,
+                merchantId=self.merchantId,
+                referenceCode=parameters['referenceCode'],
+                amount=parameters['amount'],
+                currency=parameters['currency'],
+            )
+
+        # sigantures to validate confirmation response
+        if signature_type == 2:
+            value = parameters['value']
+            last_decimal = value[-1]
+            if last_decimal == '0':
+                new_value = value[:-1]
+            else:
+                new_value = value
+
+            uncoded = "{ApiKey}~{merchantId}~{referenceSale}~{newValue}~{currency}~{statePol}".format(
+                ApiKey=self.ApiKey,
+                merchantId=self.merchantId,
+                referenceSale=parameters['reference_sale'],
+                newValue=new_value,
+                currency=parameters['currency'],
+                statePol=parameters['state_pol'],
+            )
+
+        return md5(uncoded).hexdigest()
```

Comment:
referenceSale contains the order number generated (by edx platform) to identify the basket, should not contain those characters!